### PR TITLE
WEB-7364 Fix migrator constructing TableChange when table option defi…

### PR DESCRIPTION
**Pull request by robertp**
…nitions differ

[WEB-7364](https://invoca.atlassian.net/browse/WEB-7364)

### Purpose
Fix migrator bug where strings instead of Hashes are passed to TableChange initializer

### Changes
- Add doc comment to `TableOptionsDefinition#settings`
- Use `TableOptionsDefinition#table_options` rather than `TableOptionsDefinition#settings` in `Migrator` when creating `TableChange`
- Add tests for args to TableChange spec

[WEB-7364]: https://invoca.atlassian.net/browse/WEB-7364?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by ColinDKelley — APPROVED** [review]
Looks good! Thanks for adding those specs. Just a few tiny suggestions.

File: spec/lib/declare_schema/schema_change/table_change_spec.rb [hunk 1]
```diff
@@ -7,23 +7,54 @@
...
+        subject { described_class.new(table_name, options_string, new_options) }
+
+        it 'raises ArgumentError' do
+          expect {
+            subject
+          }.to raise_exception(ArgumentError, /old_options must be a Hash but is a String: "CHARACTER SET utf8mb4 COLLATE utf8mb4_bin"/)
+        end
+      end
+
+      describe 'when new_options is a string' do
```
[the remark this answers is not in the document]
Comment (robertp) [reply spec/lib/declare_schema/schema_change/table_change_spec.rb]: f039f2adf664ddca1bdb53173b011c900d3bb00c Thanks for the suggestiosn!

File: lib/declare_schema/schema_change/table_change.rb [hunk 2]
```diff
@@ -5,8 +5,15 @@
 module DeclareSchema
   module SchemaChange
     class TableChange < Base
+
+      # @param [String] table_name  The name of the table being changed
+      # @param [Hash]   old_options The old/existing table option definitions
+      # @param [Hash]   new_options The new/existing table option definitions
       def initialize(table_name, old_options, new_options)
-        @table_name = table_name
+        old_options.is_a?(Hash) or raise ArgumentError, "old_options must be a Hash but is a #{old_options.class}: #{old_options.inspect}"
```
[the remark this answers is not in the document]
Comment (robertp) [reply lib/declare_schema/schema_change/table_change.rb]: Thanks! f039f2adf664ddca1bdb53173b011c900d3bb00c

File: Gemfile.lock [hunk 3]
```diff
@@ -1,7 +1,7 @@
 PATH
   remote: .
   specs:
-    declare_schema (1.3.1)
+    declare_schema (1.3.2.pre.1)
```
[the remark this answers is not in the document]
Comment (robertp) [reply Gemfile.lock]: Good catch! I changed version.rb to be more consistent with previous changes, as silly as that is, and forgot to re-bundle install :) f039f2adf664ddca1bdb53173b011c900d3bb00c

File: spec/lib/declare_schema/schema_change/table_change_spec.rb [hunk 4]
```diff
@@ -7,23 +7,54 @@
...
   describe '#up/down' do
-    describe '#up' do
-      it 'responds with command' do
-        statement = "ALTER TABLE #{ActiveRecord::Base.connection.quote_table_name(table_name)} CHARACTER SET utf8mb4 COLLATE utf8mb4_bin"
-        expect(subject.up).to eq("execute #{statement.inspect}\n")
+    context 'when Hashes are used to construct the TableChange' do
+      it 'does not raise ArgumentError' do
+        expect {
+          described_class.new(table_name, old_options, new_options)
+        }.not_to raise_exception(ArgumentError)
```
[the remark this answers is not in the document]
Comment (robertp) [reply spec/lib/declare_schema/schema_change/table_change_spec.rb]: f039f2adf664ddca1bdb53173b011c900d3bb00c

File: spec/lib/declare_schema/schema_change/table_change_spec.rb [hunk 5]
```diff
@@ -7,23 +7,54 @@
...
 
-    describe '#down' do
-      it 'responds with command' do
-        statement = "ALTER TABLE #{ActiveRecord::Base.connection.quote_table_name(table_name)} CHARACTER SET utf8 COLLATE utf8_ci"
-        expect(subject.down).to eq("execute #{statement.inspect}\n")
+    context 'when Strings are used to construct the TableChange' do
+      describe 'when old_options is a string' do
+        subject { described_class.new(table_name, options_string, new_options) }
+
+        it 'raises ArgumentError' do
```
[the remark this answers is not in the document]
Comment (robertp) [reply spec/lib/declare_schema/schema_change/table_change_spec.rb]: f039f2adf664ddca1bdb53173b011c900d3bb00c

File: spec/lib/declare_schema/schema_change/table_change_spec.rb [hunk 6]
```diff
@@ -7,23 +7,54 @@
...
+        end
       end
     end
 
-    describe '#down' do
-      it 'responds with command' do
-        statement = "ALTER TABLE #{ActiveRecord::Base.connection.quote_table_name(table_name)} CHARACTER SET utf8 COLLATE utf8_ci"
-        expect(subject.down).to eq("execute #{statement.inspect}\n")
+    context 'when Strings are used to construct the TableChange' do
+      describe 'when old_options is a string' do
```
[the remark this answers is not in the document]
Comment (robertp) [reply spec/lib/declare_schema/schema_change/table_change_spec.rb]: Thanks! f039f2adf664ddca1bdb53173b011c900d3bb00c

**Pull request merged**